app/frontend/helpers/printAda.ts

```diff
@@ -1,3 +1,4 @@
 // eslint-disable-next-line no-unused-vars
 import {Lovelace} from '../state'
-export default (coins: Lovelace): string => (coins * 0.000001).toFixed(6)
+export default (coins: Lovelace, decimals = 6): string =>
```

Comment:
why did you have to change this?

the new function has an off-by-one error, e.g. for `1645466` it prints `1.645465`

Reply:
ah got it - for printing the number rounded to a different amount of decimals, still, it has an off by one error

Reply:
but stll, why `(coins * 0.000001).toFixed(decimals)` would be an issue?

Reply:
you are exactly right, the reason was to not round the numbers, it was Michal's request to not round the numbers and also made sense to me not show for example `5.345` in account tile and `5.344622` in total balance just above, its pretty obvious if you have only one account. I didn't notice the 1 lovelace difference I am not sure if it happens always or just in some cases, anyway thanks a lot, even thou tx signing don't depend on this so submitting txs should be safe,  we cant display something different in UI and device. 

I dont mind reverting it and having it rounded or we can separate helper for this.

Reply:
I see there are long threads with many quite elaborate implementations tackling this for the general case, but given we are formatting ADA here we already know the maximum precision of the number and we are multiplying the amount in lovelaces by 0.000001 so there's no point in pretending this is a general purpose currency formatting function.

I would just do `toFixed(6).slice(0, -7 + decimals)` and assert on decimals being at most 6 as there is no point in formatting ADA to more than 6 decimals

Reply:
ok I would not overthink it as there are apparently floating operation edgecases in pretty much any "smart" implementation I found. Something like this would do:

```
const coinsDivided = coins * 0.000001;

if (decimals === 0) {
  return Math.floor(coinsDivided).toString() 
} else if (decimals > 0 && decimals <= 6) {
  return coinsDivided.toFixed(7).slice(0, -7 + decimals)
} else {
  return coinsDivided.toFixed(decimals)
}
```

Reply:
you a right I totally agree, thanks for the solution :+1: